### PR TITLE
bt-274: Connect Talent Onboarding to API

### LINKED ISSUE #274: Connect Talent Onboarding to API
**Acceptance criteria:**

- [x] - After clicking on the “Continue” button on the Sign up page, talent who chooses “I’m looking for a job” is redirected to the step “01/05” page of Onboarding Talent Flow.

- [x]  - When the Talent fills in all required fields on the “Profile” step and clicks on the “Next” button is redirected to the “BSA badges” step

- [x] - Talent can choose badges (“Your average project score”, “Your average lectures score”, “Level of English” is chosen by default) on the “BSA badges” step  or click on the “Next” button and talent is redirected to the “Skills and Projects” step

- [x] - When the Talent fills in all required fields on the “Skills and Projects” step and clicks on the “Next” button is redirected to the “CV and Contacts” step

- [x] - When the Talent fills in all required fields on the “CV and Contacts” step and clicks on the “Next” button is redirected to the “Preview” page

- [ ] - All the filled information should be displayed on the “Preview” page

- [x] - Talent's information should be saved in the fields when switching between steps

- [x] - When required fields are empty or invalid data is entered (does not match validation requirements), fields are highlighted in red color and a validation message is displayed, a talent stays on this page and can’t move on next step

- [x] - When the talent moves on a new step the timeline should automatically progress

### LINKED ISSUE #337: Create basic endpoint to get user details by id
Create basic endpoint for this page (Screenshot 1)
- [x] create query in repository 
- [x] create method in service
- [x] create controller for this route

Screenshot 1
Image

### PR DESCRIPTION
**Pull request by nika-sirak**
Created new migration where update columns - full_name(set nullable) and experience years(change type to decimal) and added completedStep(type enum)
Added new endpoint "/user-details/:userId", created method in service, created controller for this route
Updated user-details.repository
This task was related to [Create basic endpoint to get user details by id#337](https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/issues/337) and https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/pull/309
Updated updateTalentDetails and added getTalentDetails actions. Stored data to Redux state.

Talent's information is saved in the fields when switching between steps or when reloading page.

### CODE REVIEW

**Comment by JosuerBague** [comment]
@nika-sirak Does this branch contain changes from bt-220?

File: backend/src/bundles/user-details/user-details.controller.ts [hunk 1]
```diff
@@ -358,6 +371,46 @@ class UserDetailsController extends ControllerBase {
...
+            params: Partial<UserDetailsFindRequestDto>;
+        }>,
+    ): Promise<ApiHandlerResponse> {
+        const { userId } = options.params;
+
+        return {
+            status: HttpCode.OK,
+            payload: userId
+                ? await this.userDetailsService.findByUserId(userId)
+                : null,
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/user-details.controller.ts]: If user didn't find please throw 404 Not Found Error

File: frontend/src/app/app.tsx [hunk 2]
```diff
@@ -1,15 +1,44 @@
...
+    }));
+
+    const [isUserFullfilled, setIsUserFullfilled] = useState(false);
+
+    useEffect(() => {
+        setIsUserFullfilled(
+            dataStatus === 'fulfilled' || dataStatus == 'rejected',
+        );
+    }, [dataStatus]);
+
```
Comment (MaksymGavatuk) [inline frontend/src/app/app.tsx]: You don't need this code, unnecessary complication
```suggestion
    const isLoading = dataStatus === DataStatus.PENDING; 

```

File: backend/src/bundles/user-details/user-details.controller.ts [hunk 3]
```diff
@@ -128,6 +129,18 @@ class UserDetailsController extends ControllerBase {
...
                 ),
         });
+
+        this.addRoute({
+            path: UserDetailsApiPath.$ID,
+            method: 'GET',
+            handler: (options) => {
+                return this.findByUserId(
+                    options as ApiHandlerOptions<{
+                        params: { userId: string };
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/user-details.controller.ts]: I think we need Dto type for this
```suggestion
                        params: Dto;
```

File: frontend/src/bundles/talent-onboarding/talent-onboarding-api.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+        const response = await this.load(
+            this.getFullEndpoint(UserDetailsApiPath.ROOT, userId, {}),
+            {
+                method: 'GET',
+                contentType: ContentType.JSON,
+                hasAuth: true,
+            },
+        );
+        return await response.json<UserDetailsGeneralCustom>();
```
Comment (MaksymGavatuk) [inline frontend/src/bundles/talent-onboarding/talent-onboarding-api.ts]: Don't use return await (http://hassansin.github.io/Why-return-await-Is-a-Bad-Idea). Only in specific logic or function. Do we really need ?

File: frontend/src/bundles/talent-onboarding/talent-onboarding-api.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        const response = await this.load(
+            this.getFullEndpoint(UserDetailsApiPath.ROOT, {}),
+            {
+                method: 'POST',
+                contentType: ContentType.JSON,
+                payload: JSON.stringify(payload),
+                hasAuth: true,
+            },
+        );
+        return await response.json<UserDetailsGeneralCustom>();
```
Comment (MaksymGavatuk) [inline frontend/src/bundles/talent-onboarding/talent-onboarding-api.ts]: The same

File: frontend/src/bundles/talent-onboarding/talent-onboarding-api.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        const response = await this.load(
+            this.getFullEndpoint(UserDetailsApiPath.ROOT, {}),
+            {
+                method: 'PATCH',
+                contentType: ContentType.JSON,
+                payload: JSON.stringify(payload),
+                hasAuth: true,
+            },
+        );
+        return await response.json<UserDetailsGeneralCustom>();
```
Comment (MaksymGavatuk) [inline frontend/src/bundles/talent-onboarding/talent-onboarding-api.ts]: The same

File: frontend/src/bundles/talent-onboarding/store/actions.ts [hunk 7]
```diff
@@ -9,8 +9,32 @@ const updateTalentDetails = createAsyncThunk<
...
+
+    if ('badges' in registerPayload) {
+        return registerPayload;
+    }
+
+    const userDetails = await talentOnBoardingApi.getUserDetailsByUserId({
+        userId: registerPayload.userId,
+    });
+
+    return await (userDetails
```
Comment (MaksymGavatuk) [inline frontend/src/bundles/talent-onboarding/store/actions.ts]: The same return await

File: frontend/src/bundles/talent-onboarding/store/actions.ts [hunk 8]
```diff
@@ -9,8 +9,32 @@ const updateTalentDetails = createAsyncThunk<
...
+});
+
+const getTalentDetails = createAsyncThunk<
+    UserDetailsGeneralCustom | null,
+    UserDetailsGeneralCustom,
+    AsyncThunkConfig
+>(`${sliceName}/get-talent-details`, async (findPayload, { extra }) => {
+    const { talentOnBoardingApi } = extra;
+
+    return await talentOnBoardingApi.getUserDetailsByUserId({
```
Comment (MaksymGavatuk) [inline frontend/src/bundles/talent-onboarding/store/actions.ts]: The same return await

**Review by MaksymGavatuk — CHANGES_REQUESTED** [review]

File: frontend/src/app/app.tsx (hunk 2 above)
Comment (nika-sirak) [reply frontend/src/app/app.tsx]: Could I leave this code? Dasha is working on it. I will merge updates after she finishes that part.
https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/pull/309

File: frontend/src/bundles/talent-onboarding/store/actions.ts [hunk 9]
```diff
@@ -9,8 +9,32 @@ const updateTalentDetails = createAsyncThunk<
...
->(`${sliceName}/update-talent-details`, (registerPayload) => {
-    return registerPayload;
+>(
+    `${sliceName}/update-talent-details`,
+    (registerPayload, { getState, extra }) => {
+        const { talentOnBoardingApi } = extra;
+
+        if ('badges' in registerPayload) {
+            return registerPayload;
+        }
```
Comment (nikita-remeslov) [inline frontend/src/bundles/talent-onboarding/store/actions.ts]: why do we need this logic?

File: frontend/src/bundles/talent-onboarding/store/actions.ts [hunk 10]
```diff
@@ -9,8 +9,32 @@ const updateTalentDetails = createAsyncThunk<
...
+        const { talentOnBoarding } = getState();
+
+        return talentOnBoarding.completedStep
+            ? talentOnBoardingApi.updateUserDetails(registerPayload)
+            : talentOnBoardingApi.createUserDetails(registerPayload);
```
Comment (nikita-remeslov) [inline frontend/src/bundles/talent-onboarding/store/actions.ts]: create 3 action

1. POST
2. PATCH
3. combine logic where you should dispatch POST or PATCH

File: frontend/src/bundles/talent-onboarding/store/actions.ts [hunk 11]
```diff
@@ -9,8 +9,32 @@ const updateTalentDetails = createAsyncThunk<
     UserDetailsGeneralCustom,
     UserDetailsGeneralCustom,
     AsyncThunkConfig
->(`${sliceName}/update-talent-details`, (registerPayload) => {
-    return registerPayload;
+>(
+    `${sliceName}/update-talent-details`,
+    (registerPayload, { getState, extra }) => {
```
Comment (nikita-remeslov) [inline frontend/src/bundles/talent-onboarding/store/actions.ts]:
```suggestion
    (registerPayload, { getState, extra, dispatch }) => {
```

File: frontend/src/bundles/talent-onboarding/store/actions.ts (hunk 10 above)
Comment (nikita-remeslov) [inline frontend/src/bundles/talent-onboarding/store/actions.ts]:
```suggestion
        return talentOnBoarding.completedStep
            ? dispatch(patchAction(registerPayload))
            : dispatch(postAction(registerPayload))
```

File: frontend/src/bundles/talent-onboarding/store/slice.ts [hunk 12]
```diff
@@ -27,11 +34,41 @@ const { reducer, actions, name } = createSlice({
...
+        builder.addCase(getTalentDetails.fulfilled, (state, action) => {
+            state.dataStatus = DataStatus.FULFILLED;
+            for (const key in action.payload) {
+                const typedKey = key as keyof UserDetailsFindRequestDto;
+
+                state[typedKey] =
+                    typedKey === 'englishLevel' &&
+                    action.payload[typedKey] === null
+                        ? ' '
+                        : action.payload[typedKey];
```
Comment (nikita-remeslov) [inline frontend/src/bundles/talent-onboarding/store/slice.ts]: why do we need this logic?

File: frontend/src/bundles/talent-onboarding/store/slice.ts [hunk 13]
```diff
@@ -27,11 +34,41 @@ const { reducer, actions, name } = createSlice({
...
         builder.addCase(updateTalentDetails.fulfilled, (state, action) => {
+            state.dataStatus = DataStatus.FULFILLED;
+
             for (const key in action.payload) {
                 const typedKey = key as keyof UserDetailsUpdateRequestDto;
-                state[typedKey] = action.payload[typedKey];
+                state[typedKey] =
+                    typedKey === 'englishLevel' &&
+                    action.payload[typedKey] === null
+                        ? ' '
```
Comment (nikita-remeslov) [inline frontend/src/bundles/talent-onboarding/store/slice.ts]: " " - why do we need it?

File: frontend/src/bundles/talent-onboarding/talent-onboarding-api.ts [hunk 14]
```diff
@@ -0,0 +1,71 @@
+import {
+    ApiPath,
+    ContentType,
+    UserDetailsApiPath,
+} from 'shared/build/index.js';
```
Comment (nikita-remeslov) [inline frontend/src/bundles/talent-onboarding/talent-onboarding-api.ts]: we don't import from shared directly

File: shared/src/bundles/user-details/enums/completed-step.enum.ts [hunk 15]
```diff
@@ -0,0 +1,9 @@
+const CompletedStep = {
```
Comment (nikita-remeslov) [inline shared/src/bundles/user-details/enums/completed-step.enum.ts]:
```suggestion
const OnboardingSteps = {
```

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]

File: frontend/src/bundles/talent-onboarding/store/actions.ts (hunk 9 above)
Comment (nika-sirak) [reply frontend/src/bundles/talent-onboarding/store/actions.ts]: I forgot to write a comment that it's a temporary solution.

**Review by nikita-remeslov — APPROVED** [review]

File: frontend/src/app/app.tsx [hunk 16]
```diff
@@ -22,14 +22,14 @@ const App: React.FC = () => {
...
-    if (dataStatus == DataStatus.PENDING) {
+    if (dataStatus === DataStatus.PENDING) {
         return <Loader />;
     }
 
     return (
         <>
             <RouterOutlet />
-            <Notifications />
+            {dataStatus === DataStatus.REJECTED && <Notifications />}
```
Comment (nikita-remeslov) [inline frontend/src/app/app.tsx]: ?

Comment (anton-otroshchenko) [inline frontend/src/app/app.tsx]:
```suggestion
           <Notifications />
```

**Review by anton-otroshchenko — CHANGES_REQUESTED** [review]

File: frontend/src/app/app.tsx (hunk 16 above)
Comment (nika-sirak) [reply frontend/src/app/app.tsx]: Fixed it

Comment (nika-sirak) [reply frontend/src/app/app.tsx]: Fixed it

**Review by anton-otroshchenko — APPROVED** [review]

**Review by savinaDaria — APPROVED** [review]

**Pull request merged**